src/Agent.Worker/TaskManager.cs

```diff
@@ -128,23 +129,71 @@ private async Task DownloadAsync(IExecutionContext executionContext, Pipelines.T
             // the task zip is large. It would be nice to print the localized name, but it is not
             // available from the reference included in the job message.
             executionContext.Output(StringUtil.Loc("DownloadingTask0", task.Name));
-            string zipFile;
+            string zipFile = string.Empty;
             var version = new TaskVersion(task.Version);
 
             //download and extract task in a temp folder and rename it on success
             string tempDirectory = Path.Combine(HostContext.GetDirectory(WellKnownDirectory.Tasks), "_temp_" + Guid.NewGuid());
             try
             {
                 Directory.CreateDirectory(tempDirectory);
-                zipFile = Path.Combine(tempDirectory, string.Format("{0}.zip", Guid.NewGuid()));
-                //open zip stream in async mode
-                using (FileStream fs = new FileStream(zipFile, FileMode.Create, FileAccess.Write, FileShare.None, bufferSize: 4096, useAsync: true))
+                int retryCount = 0;
+
+                // Allow up to 10*60s for any task to be downloaded from service
+                // Timeout limit can be overwrite by environment variable
+                if (!int.TryParse(Environment.GetEnvironmentVariable("VSTS_TASK_DOWNLOAD_TIMEOUT") ?? string.Empty, out int timeoutSeconds))
+                {
+                    timeoutSeconds = 600;
+                }
+
+                while (retryCount < 3)
```

Comment:
How can we L0 test the retry and backoff logic to ensure there's not a bug?  Ideally, we could mock the work and the backoff.  Might be related to point on retry manager.

Reply:
i added `RetryNetworkException` and `RetryStreamException` L0s to make sure try works.